controllers/user.go

```diff
@@ -0,0 +1,144 @@
+	CreatedAt time.Time `json:"created_at"`
+}
+
+type LoginRequest struct {
+	Email    string `json:"email" validate:"required,email"`
+	Password string `json:"password" validate:"required,min=8"`
+}
+
+func RegisterUser(w http.ResponseWriter, r *http.Request, userService services.UserService) {
+	var requestBody RegisterRequest
+
+	err := json.NewDecoder(r.Body).Decode(&requestBody)
+	if err != nil {
+		log.Println(err)
+		http.Error(w, "Invalid request body", http.StatusBadRequest)
+		return
+	}
+
+	validate := validator.New()
+	err = validate.Struct(requestBody)
+	if err != nil {
+		log.Println(err)
+		http.Error(w, "Validation failed", http.StatusBadRequest)
+		return
+	}
+
+	newUser, err := userService.RegisterUser(requestBody.Username, requestBody.Email, requestBody.Password)
+	if err != nil {
+		switch err.Error() {
+		case "email already exists":
+			http.Error(w, "Email already exists", http.StatusConflict)
+		case "username already exists":
+			http.Error(w, "Username already taken", http.StatusConflict)
+		default:
+			log.Fatal(err)
+			http.Error(w, "Internal server error", http.StatusInternalServerError)
+		}
+
+		return
+	}
+
+	response := RegisterResponse{
+		ID:        newUser.ID,
+		Username:  newUser.Username,
+		Email:     newUser.Email,
+		CreatedAt: newUser.CreatedAt,
+	}
+
+	jsonResponse, err := json.Marshal(response)
+
+	if err != nil {
+		http.Error(w, "Internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusCreated)
+	w.Write(jsonResponse)
+}
+
+func LoginUser(w http.ResponseWriter, r *http.Request, userService services.UserService) {
+	log.Println("inside login user")
+	var requestBody LoginRequest
+
+	err := json.NewDecoder(r.Body).Decode(&requestBody)
+	if err != nil {
+		log.Println(err)
+		http.Error(w, "Invalid request body", http.StatusBadRequest)
+		return
+	}
+
+	log.Println("request body decoded")
+
+	validate := validator.New()
+	err = validate.Struct(requestBody)
+	if err != nil {
+		log.Println(err)
+		http.Error(w, "Validation failed", http.StatusBadRequest)
+		return
+	}
+
+	log.Println("request body validated")
+
+	log.Println(requestBody)
+
+	user, err := userService.AuthenticateUser(requestBody.Email, requestBody.Password)
+	if err != nil {
+		switch err.Error() {
+		case "invalid credentials":
+			http.Error(w, "Invalid email or password", http.StatusUnauthorized)
+		case "user does not exist":
+			http.Error(w, "User does not exist", http.StatusNotFound)
+		default:
+			http.Error(w, "Internal server error", http.StatusInternalServerError)
+		}
+		return
+	}
+
+	// Generate JWT token
+	token, err := utils.GenerateJWT(user.ID, user.Username)
+	if err != nil {
+		http.Error(w, "Failed to generate token", http.StatusInternalServerError)
+		return
+	}
+
+	// Set the JWT token in the response
+	http.SetCookie(w, &http.Cookie{
+		Name:     "auth_token",
+		Value:    token,
+		HttpOnly: true,
+		Secure:   true,
```

Comment:
secure true must only be in production env and not in development env

Reply:
I read up upon this, not required, let me explain.

If you see, I had the `Secure` flag set to `true` while working on the PR.

Now, if you read upon the [docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#block_access_to_your_cookies:~:text=A%20cookie%20with,modify%20the%20information.) for this, you will see this:

> A cookie with the Secure attribute is only sent to the server with an encrypted request over the HTTPS protocol. It's never sent with unsecured HTTP (except on localhost), which means [man-in-the-middle](https://developer.mozilla.org/en-US/docs/Glossary/MitM) attackers can't access it easily. Insecure sites (with http: in the URL) can't set cookies with the Secure attribute.

Now, if that is the case, how come it worked on local where I don't have `https`?

Well, it turns out that the browser does send the cookie, if I am running our application locally without HTTPS. There are reasons to it:

- Modern browsers treat all requests to localhost or 127.0.0.1 as secure, regardless of whether they're actually encrypted or not.
- When you access your local server, it's treated as a separate origin from other websites you visit.
- Due to this special treatment of localhost, the browser sends the cookie even if it has the Secure attribute set.
- Treating localhost as secure allows for easier testing of HTTPS features.
- It helps ensure that local development behaves similarly to production environments.
- It's designed to make local development easier while maintaining consistency with how cookies behave in production.